src/simmate/apps/evolution/stop_conditions/stop_conditions.py

```diff
@@ -21,35 +27,114 @@ def check(self, search):
             return False
-        min_energy = min(energies)
 
-        # look at the structure energies and see what the lowest is
-        if min_energy <= self.energy_limit:
-            logging.info("\nEnergy limit hit. Stopping Search")
+        # count the number of new individuals added AFTER the best one. If it is
+        # more than best_survival_cutoff, we stop the search.
+        # Note, we look at all structures that have an energy_per_atom greater
+        # than 1meV/atom higher than the best structure. The +1meV ensures
+        # we aren't prolonging the calculation for insignificant changes in
+        # the best structure. Looking at the energy also ensures that we are
+        # only counting completed calculations.
+        # BUG: this filter needs to be updated to fitness_value and not
+        # assume we are using energy_per_atom
+        num_new_structures_since_best = search.individuals.filter(
+            energy_per_atom__gt=best.energy_per_atom + search.convergence_cutoff,
+            finished_at__gte=best.finished_at,
+        ).count()
+        if num_new_structures_since_best > search.best_survival_cutoff:
+            logging.info(
+                "Best individual has not changed after "
+                f"{self.best_survival_cutoff} new individuals added."
+            )
             return True
+        # If we reached this point, then we haven't hit a stop condition yet!
+        return False
 
-        # look how long the minimum structure has been there #!!! ERROR IN LOGIC
-        # min_index = energies.index(min_energy)
-        # if len(energies) - min_index >= self.same_min_structures:
-        #     print('\nMin-structure reign limit hit. Stopping Search')
-        #     return True
 
-        # If no conditions above are met, return False to continue the search
-        return False
+class ExpectedStructure(StopCondition):
+    """
+    Stops the search if the provided expected structure has been found.
+    """
+
+    def check(self, search):
+        # !!! Here we compare individuals using pymatgen's StructureMatcher. An
+        # alternative would be to use the searches validator property like we
+        # do to find unique individuals in the search itself. Depending on how
+        # slow the following for loop is, that may be necessary
+        # get expected structure and make sure it is a pymatgen Structure object
+        expected_structure_dynamic = search.expected_structure
+        # make sure we have an expected structure. If not we won't use this stop condition
+        if not expected_structure_dynamic:
+            logging.warning(
+                "No expected_structure was provided. The ExpectedStructure stop condition will be ignored"
+            )
+            return False
+        expected_structure = Structure.from_dynamic(expected_structure_dynamic)
+        # get completed individuals
+        individuals = search.individuals_completed.order_by("finished_at").all()
```

Comment:
yeah this is a bottleneck and wasn't an issue when I ran these checks manually. One workaround is caching the last check (as a class attribute, not in the database), and then doing `.individuals_completed.filter(finished_at__gte=last_check_timestamp)`  to only check the new ones